keras_core/distribution/distribution_lib.py

```diff
@@ -0,0 +1,142 @@
+        shape,
+        axis_names,
+        devices=None,
+    ):
+        """Initialize the DeviceMesh for the given topology.
+
+        Args:
+            shape: tuple of list of integers. The shape of the overall
+                DeviceMesh, e.g. `(8,)` for a data parallel only distribution,
+                or `(4, 2)` for a model+data parallel distribution.
+            axis_names: List of string. The logical name of the each axis for
+                the DeviceMesh. The length of the `axis_names` should match to
+                the rank of the `shape`. The `axis_names` will be used to
+                match/create the `TensorLayout` when distribute the data and
+                weights.
+            devices: Optional list of devices. Default to all the available
+                devices locally from `list_devices()`.
+        """
+        pass
+
+
+class TensorLayout:
+    """The layout of a Tensor.
+
+    This is aligned with `jax.sharding.NamedSharding` and `tf.dtensor.Layout`,
+    which allocate the tensor to its logic axis based on the `DeviceMesh`. With
+    `DeviceMesh` and `TensorLayout`, the actual mapping between a Tensor to the
+    physical devices can be determined.
+
+    See more details in [jax.sharding.NamedSharding]
+    (https://jax.readthedocs.io/en/latest/jax.sharding.html#jax.sharding.NamedSharding)
+    and [tf.dtensor.Layout]
+    (https://www.tensorflow.org/api_docs/python/tf/experimental/dtensor/Layout).
+    """
+
+    def __init__(self, axes):
+        """Initialize the TensorLayout with axis names.
+
+        Args:
+            axes: list of strings that should map to the `axis_names` in
+                `DeviceMesh`. For any dimentions that doesn't need any sharding,
+                A `None` can be used a placeholder.
+        """
+        pass
+
+
+class Distribution:
+    """Base class for the distribution.
+
+    The `Distribution` has following key functionalities.
+
+    1. Distribute the model variables to the `DeviceMesh`.
+    2. Distribute the input data to the `DeviceMesh`.
+
+    It can create a context scope so that the framework to properly detect the
+    `Distribution` and distribute the variable/data accordingly.
+    """
+
+    def __init__(self, device_mesh):
+        pass
+
+    def distribute_data(self, data):
+        """Shard the input data based on the Distribution setting.
+
+        Args:
+            data: input data, can be Tensor and np.Array.
+
+        Returns:
+            distributed data tensor.
+        """
+        pass
+
+    def distribute_variable(self, variable):
+        """Distribute the variable based on the Distribution setting.
+
+        Args:
+            variable: a keras variable for distribution.
+
+        return:
+            distributed keras variable.
+        """
+        pass
+
+    @contextlib.contextmanager
+    def scope(self):
+        """Context manager to make the `Distribution` current."""
+        pass
+
+
+def get_distribution():
```

Comment:
Just `distribution()`. We have `dtype_policy()` and `image_data_format()` and `floatx()` similarly. It would be clearer to have a `set_` prefix for these, but that's the convention we have today.

Reply:
Done.